feat: Add exchange rate table and dictionary

rafaeelaudibert opened the pull request:
This will be used by Web Analytics to convert revenue between currencies. There's a `historical.csv` dictionary that we'll use to backfill everything since January 1st 2000 all the way to the end of 2024. Everything after 2024 will be generated via a Dagster job, it'll help us guarantee that this is working correctly.

Not every value is filled in, this is a sparse table, implying that you need to look at the closest date smaller than your current date if you wanna know the current rate. We'll have a hogQL function that will handle that. This is not only required for historical data, but also in case we're in a new date and we haven't gotten any rates from the new day yet.

rafaeelaudibert committed:
feat: Add exchange rate table and dictionary

This will be used by Web Analytics to convert revenue between currencies. There's a `historical.csv` dictionary that we'll use to backfill everything since January 1st 2000 all the way to the end of 2024. Everything after 2024 will be generated via a Dagster job, it'll help us guarantee that this is working correctly.

Not every value is filled in, this is a sparse table, implying that you need to look at the closest date smaller than your current date if you wanna know the current rate. We'll have a hogQL function that will handle that. This is not only required for historical data, but also in case we're in a new date and we haven't gotten any rates from the new day yet.

rafaeelaudibert commented on posthog/models/exchange_rate/sql.py:
```diff
@@ -0,0 +1,161 @@
+        engine=MergeTreeEngine("exchange_rate", replication_scheme=ReplicationScheme.REPLICATED),
+        on_cluster_clause=ON_CLUSTER_CLAUSE(on_cluster),
+    )
+)
```
@PostHog/clickhouse I've copied most of this from @robbie-c's `channel_definition` table. I'm not really sure the `ON CLUSTER` calls here are correct, can you check that for me, please?

Daesgar replied:
Yeah this is fine 👍 by default you'll use the on cluster clause except when running on coordinator nodes

rafaeelaudibert commented on posthog/models/exchange_rate/sql.py:
```diff
@@ -0,0 +1,166 @@
+ORDER BY (date, currency);
+""".format(
+        table_name=EXCHANGE_RATE_TABLE_NAME,
+        engine=MergeTreeEngine(
+            "exchange_rate", replication_scheme=ReplicationScheme.REPLICATED, replacing_col="version"
+        ),
```
@PostHog/clickhouse Is this accurate? We'll store several of these rows during the day, once every hour, this way we'll keep up-to-date values from the last day. We'll then run a background job at night to confirm the value from the last day. The idea is that I want to keep the latest value. See the dictionary creation below as well where I need to query on `anyLast` to get the right value inside the dictionary.

rafaeelaudibert replied:
Solved below, using ReplicatingMergeTree instead

orian commented on posthog/models/exchange_rate/sql.py:
```diff
@@ -0,0 +1,166 @@
+                continue
+
+    # Fill in missing dates
+    start_date = datetime.datetime(2000, 1, 1)
```
Just for a clarification, the exchange rates were published for what time zone? UTC or some US one?

rafaeelaudibert replied:
The idea here is to store it in UTC at all times. I'll double-check the file we downloaded is using UTC timezones. It's not the most accurate file, but I don't expect us to *need* extremely accurate historical data, rates won't change much between midnight UTC and midnight PST/EST.

rafaeelaudibert replied:
I can actually remove this `"Fill in missing dates"` because it'll be empty anyway, it was helpful for debugging, but now that this is working, we don't need it anymore

orian replied:
AFAIK the rates usually are there only when official market is running, and official rates usually are what national banks are publishing the day after.

orian commented on posthog/models/exchange_rate/sql.py:
```diff
@@ -0,0 +1,166 @@
+ORDER BY (date, currency);
+""".format(
+        table_name=EXCHANGE_RATE_TABLE_NAME,
+        engine=MergeTreeEngine(
```
I think you should use ReplacingMergeTree here with ver instead of replacing_col.

orian commented on posthog/models/exchange_rate/sql.py:
```diff
@@ -0,0 +1,166 @@
+INSERT INTO exchange_rate (date, currency, rate) VALUES
+{
+''',
+'''.join(f"(toDate('{date}'), '{currency}', {rate})" for date, currency, rate in exchange_rates)},
```
@PostHog/clickhouse this is worth to take a look at, not sure how we use to put fixtures for ClickHouse, but because each test run may need to run it, it's worth to be done right.

rafaeelaudibert commented on posthog/models/exchange_rate/historical.csv:
I initially removed USD from this list because it's trivial, but it might be a good idea to have it in the list because it'll allow us to not have an extra check if someone wants to convert from USD to a different currency (Europeans might want that, for example)

orian commented on posthog/models/exchange_rate/sql.py:
```diff
@@ -0,0 +1,166 @@
+    rate Decimal64(4)
+)
+PRIMARY KEY (date, currency)
+SOURCE(CLICKHOUSE(TABLE '(SELECT date, currency, anyLast(rate) AS rate FROM {EXCHANGE_RATE_TABLE_NAME} GROUP BY date, currency)' PASSWORD '{CLICKHOUSE_PASSWORD}'))
```
why passing a password here is necessary?

Daesgar replied:
ClickHouse requires it, otherwise it will fail, even though it's querying itself... 🤷

Daesgar replied:
Also I think you want to use QUERY instead of table here. And I'd say that parenthesis are not needed:

```suggestion
SOURCE(CLICKHOUSE(QUERY 'SELECT date, currency, anyLast(rate) AS rate FROM {EXCHANGE_RATE_TABLE_NAME} GROUP BY date, currency' PASSWORD '{CLICKHOUSE_PASSWORD}'))
```

rafaeelaudibert replied:
Oh, you're right, changing to QUERY

rafaeelaudibert commented on posthog/models/exchange_rate/sql.py:
Rather than blindly including all of the existing currencies, I'll actually create a new constant in this file listing all of the supported currencies - which will be all of the ones returned by OER. This will be helpful in https://github.com/PostHog/posthog/pull/29323 as well because it'll allow us to ignore the rates we don't care about

Daesgar commented on posthog/models/exchange_rate/sql.py:
```diff
@@ -0,0 +1,166 @@
+)
+
+
+def EXCHANGE_RATE_DATA_BACKFILL_SQL(exchange_rates=None):
+    if exchange_rates is None:
+        exchange_rates = HISTORICAL_EXCHANGE_RATE_TUPLES()
+
+    return f"""
+INSERT INTO exchange_rate (date, currency, rate) VALUES
+{
```
Can't we perform this historical backfill in Dagster instead? We can create the schemas here and delegate to the Dagster job the historical and successive loads. Unless you need to have the data right after running the migration, but since you mentioned you were going to use Dagster to update this, it may be worth performing all the data load there and avoid storing a CSV here.

Nothing blocking, just a suggestion in case it makes sense for your use case!

rafaeelaudibert replied:
We could, but I thought a one-off thing would make more sense if done via a migration than a Dagster job. This is just 8MB. Happy to change, though. I can store this on S3 somewhere and query it from Dagster, insert it, and then remove that job, it's just a huge workaround

rafaeelaudibert replied:
@Daesgar Something else crossed my mind, we'll want *some* currencies locally, so I'd argue we should keep this here. The CSV is pretty small.

Daesgar replied:
Yeah completely it was just a suggestion in case it made sense! Totally fine

Daesgar commented on posthog/clickhouse/migrations/0101_add_exchange_rates.py:
```diff
@@ -0,0 +1,18 @@
+)
+
+operations = [
+    # Drop tables to allow this to rerun
+    run_sql_with_exceptions(DROP_EXCHANGE_RATE_TABLE_SQL),
+    run_sql_with_exceptions(DROP_EXCHANGE_RATE_DICTIONARY_SQL),
+    # Recreate them all
+    run_sql_with_exceptions(EXCHANGE_RATE_TABLE_SQL()),
+    run_sql_with_exceptions(EXCHANGE_RATE_DATA_BACKFILL_SQL()),
+    run_sql_with_exceptions(EXCHANGE_RATE_DICTIONARY_SQL()),
```
Adding the SQL for the coordinators as well 🙂 

```suggestion
    # Drop tables to allow this to rerun
    run_sql_with_exceptions(DROP_EXCHANGE_RATE_TABLE_SQL),
    run_sql_with_exceptions(DROP_EXCHANGE_RATE_TABLE_SQL(on_cluster=False), node_role=NodeRole.COORDINATOR),
    run_sql_with_exceptions(DROP_EXCHANGE_RATE_DICTIONARY_SQL),
    run_sql_with_exceptions(DROP_EXCHANGE_RATE_DICTIONARY_SQL(on_cluster=False), node_role=NodeRole.COORDINATOR),
    # Recreate them all
    run_sql_with_exceptions(EXCHANGE_RATE_TABLE_SQL()),
    run_sql_with_exceptions(EXCHANGE_RATE_TABLE_SQL(on_cluster=False), node_role=NodeRole.COORDINATOR),
    run_sql_with_exceptions(EXCHANGE_RATE_DATA_BACKFILL_SQL()),
    run_sql_with_exceptions(EXCHANGE_RATE_DATA_BACKFILL_SQL(on_cluster=False), node_role=NodeRole.COORDINATOR),
    run_sql_with_exceptions(EXCHANGE_RATE_DICTIONARY_SQL()),
    run_sql_with_exceptions(EXCHANGE_RATE_DICTIONARY_SQL(on_cluster=False), node_role=NodeRole.COORDINATOR),
```

rafaeelaudibert committed:
feat: Add USD back

It's useful to store USD = 1 because it'll allow us to let people use different base currencies even if they're ingesting stuff in USD

rafaeelaudibert committed:
refactor: Small fixes to Clickhouse code + new currencies list

rafaeelaudibert committed:
feat: Add BTC rates

They suck because they're pretty small compared to the BTC, so we'll need to increase our decimals to 10 places

rafaeelaudibert committed:
fix: Make all constants a function inside sql.py

rafaeelaudibert committed:
fix: Stop passing on_cluster=false for INSERT query

It's inserted wherever it is, so we don't need to specify this here

rafaeelaudibert committed:
fix: Remove Zimbabwean dollar

Hyperinflation makes the number huge, let's remove it, we can figure it
out if we ever have to support it - unlikely

rafaeelaudibert committed:
refactor: Use a `RANGE_HASHED` instead of a `COMPLEX_KEY_HASHED` dict

Rationale is written in the comments

rafaeelaudibert committed:
fix: Add missing RANGE clause

rafaeelaudibert committed:
fix: Remove trailing comma
